src/resources/common.ts

```diff
@@ -1,31 +1,97 @@
+import {
+  Resource as ProtoResource,
+  ResourceType,
+  PolicyResource,
+  ResourceDeclareRequest,
+  ResourceDeclareResponse,
+  ActionMap,
+} from '@nitric/api/proto/resource/v1/resource_pb';
+import resourceClient from './client';
+
+export type Permission = 'reading' | 'writing' | 'deleting';
+
+const everything: Permission[] = ['reading', 'writing', 'deleting'];
+
+export type ResourcePermMap = Record<Permission, ActionMap[keyof ActionMap][]>;
+
+type ActionsList = ActionMap[keyof ActionMap][];
+
+export const permsToActions = (
+  resourcePermMap: ResourcePermMap,
+  ...perms: Permission[]
+): ActionsList => {
+  return perms.reduce((actions, perm) => {
+    if (!['deleting', 'reading', 'writing'].includes(perm)) {
+      throw new Error(
+        `unknown bucket permission ${perm}, supported permissions are ${everything.join(
+          ', '
+        )}`
+      );
+    }
+
+    return [...actions, ...resourcePermMap[perm]];
+  }, []);
+};
```

Comment:
Not all resource types will use bucket permissions. This code cannot be re-used. Each resource implementation would need to provide its on perms -> actions resolver.